ebpf/session.go

```diff
@@ -357,45 +499,250 @@ func (s *session) walkStack(sb *stackBuilder, stack []byte, pid uint32, stats *s
 	}
 }
 
-func (s *session) updateSampleRate(sampleRate int) error {
-	if s.options.SampleRate == sampleRate {
-		return nil
+func (s *session) readEvents(events *perf.Reader,
```

Comment:
There's no test for session.go and I think it would be great to think about it since you have many locks and goroutine, this means most likely add a interface for the perf package used.

Just to make sure it calls correctly perf and doesn't leak and has a race.

Reply:
totally agree, I want this a lot, however mocking out the whole cilium/ebpf library is not trivial, there's a lot of structures used, like maps, programs, perf reader.  Lets create an issue and improve later

Reply:
https://github.com/grafana/pyroscope/issues/2576